src/build.jl

```diff
@@ -65,6 +65,20 @@ function build_folder(folder)
     write(path, log)
 end
 
+"""
+    safe_instantiate(folders)
+
+Install all packages sequentially since Pkg.jl is thread-unsafe.
+See https://github.com/JuliaLang/Pkg.jl/issues/2219 for details.
+"""
+function safe_instantiate(folders)
```

Comment:
The problem with this approach is that you end up with the last activated environment when you run `safe_instantiate` or `parallel_build`, i.e., it will affect your project environment when you run it from the REPL. Additionally, in my experience it is quite messy and easily leads to broken/incorrectly loaded environments if one activates and instantiates multiple environments in one session (but maybe it's less of an issue if no packages are loaded, I am not sure how problematic precompilation is).

Instead one could use the same approach as in https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/blob/671f960726b0b9826b9e5a81cc3e55d6c4ba0912/docs/make.jl#L10-L34 (and similarly in AbstractGPs). BTW why does TuringTutorials use multiple threads instead of running multiple system commands asynchronously? Are there any advantages?

Reply:
> The problem with this approach is that you end up with the last activated environment when you run safe_instantiate or parallel_build, i.e., it will affect your project environment when you run it from the REPL. Additionally, in my experience it is quite messy and easily leads to broken/incorrectly loaded environments if one activates and instantiates multiple environments in one session (but maybe it's less of an issue if no packages are loaded, I am not sure how problematic precompilation is).

Ah okay. Yes. Good point, that could indeed lead to issues. I could start a new Julia instance for each activate and instantiate. Would you consider that as a solution?

> Instead one could use the same approach as in https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/blob/671f960726b0b9826b9e5a81cc3e55d6c4ba0912/docs/make.jl#L10-L34 (and similarly in AbstractGPs). BTW why does TuringTutorials use multiple threads instead of running multiple system commands asynchronously? Are there any advantages?

In essence, we're doing the same thing, right? I just copied the `$(Base.julia_cmd()) ...` from there. The reason that I use threads is because it explicitly states (to the reader) that it runs in parallel. With `map` and `process`, it's a bit more implicit.

Reply:
>  I could start a new Julia instance for each activate and instantiate. Would you consider that as a solution?

That's what we do in the code I linked 🙂 It seems to work so I guess it should be fine.

Reply:
:+1: Now I see.

Reply:
> I just copied the $(Base.julia_cmd()) ... from there. The reason that I use threads is because it explicitly states (to the reader) that it runs in parallel. With map and process, it's a bit more implicit.

But it requires the user to set JULIA_NUM_THREADS, otherwise it won't run in parallel, will it? Whereas in the other case it just starts multiple system process in parallel and hence the system scheduler will take care of everything.

Reply:
Aha, okay. And then the new Julia instances use multiple threads because the OS gives them a new thread if needed. So, you spawn all the Julia instances at the same time? Will this not actually slow down the execution when you start many tutorials. In the case of TuringTutorials, we have 10 tutorials and some of them start `n` threads. On the GitHub CI, I set `JULIA_NUM_THREADS` to 2, so now we end up with at most 4 threads running at the same time (with the GitHub runners having 2 threads). If I would use map and spawn all tutorials at the same time, we would end up with somewhere between 10 and 20 threads (10 tutorials * 1.5 threads per tutorial on average).

Could you open an issue if you want TuringTutorials to remove the `Threads.@threads`? I think that it is out of scope for this PR, because it needs some proper testing when I change it.